### PR TITLE
Enable Dia 1.6B Model to Run on Apple Silicon GPUs (MPS Backend) (#197)

### LINKED ISSUE #197: Fails on Apple Silicon GPUs (MPS Backend) Due to Attention Shape Mismatch
## Problem
The Dia 1.6B model crashes on Apple Silicon (M-series) GPUs with the MPS backend, throwing this error:

```bash
Using device: mps
generate: starting generation loop
loc("mps_matmul"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/79ef05cb-ffe7-11ef-b000-f2a857e00a32/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":43:0)): error: incompatible dimensions
loc("mps_matmul"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/79ef05cb-ffe7-11ef-b000-f2a857e00a32/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":43:0)): error: invalid shape
LLVM ERROR: Failed to infer result type(s).
[1]    17087 abort      python example/simple-mac.py
```

This happens because of a shape mismatch in the attention mechanism when Grouped Query Attention (GQA) is enabled. The MPS backend doesn’t handle tensor broadcasting as flexibly as CUDA or CPU, causing matrix multiplication to fail.

## Root Cause
In `dia/layers.py`, the `Attention.forward` method uses `F.scaled_dot_product_attention`. With GQA enabled, the query, key, and value tensors have incompatible head dimensions, which MPS rejects.

## Proposed Fix
Replace `F.scaled_dot_product_attention` with a custom function that manually aligns the key and value tensor shapes for GQA, using `torch.matmul` to compute attention. This ensures MPS compatibility.

## Steps to Reproduce
1. Install the Dia package `pip install git+https://github.com/nari-labs/dia.git`.
2. Run the following example on an M-series Mac with MPS enabled.
```python
import torch
from dia.model import Dia

# Select device: MPS if available, else CPU
device = torch.device("mps") if torch.backends.mps.is_available() else torch.device("cpu")
print(f"Using device: {device}")

model = Dia.from_pretrained("nari-labs/Dia-1.6B", compute_dtype="float32", device=device)

text = "[S1] Dia is an open weights text to dialogue model. [S2] You get full control over scripts and voices. [S1] Wow. Amazing. (laughs) [S2] Try it now on Git hub or Hugging Face."

output = model.generate(text, use_torch_compile=False, verbose=True)

model.save_audio("simple.mp3", output)
```
3. Note the `mps_matmul` error during generation.
## Impact
This bug blocks Apple Silicon GPU users from running the model, reducing accessibility. A fix will be submitted via a pull request linked to this issue.

### PR DESCRIPTION
**Pull request by V12Hero**
**Fixes #197**

## Summary
This pull request resolves issue #197 by enabling the Dia 1.6B text-to-speech (TTS) model to run on Apple Silicon (M-series) GPUs using the Metal Performance Shaders (MPS) backend. The fix addresses a shape incompatibility in the attention mechanism that caused errors on MPS. A custom scaled dot-product attention function replaces the original implementation, ensuring compatibility and efficient execution on M-series Macs.

---

## Changes
The key modification is in `dia/layers.py`, where we replaced `F.scaled_dot_product_attention` with a custom implementation:
- **New Function**: `custom_scaled_dot_product_attention`
  - Explicitly repeats key and value tensors to match query heads for Grouped Query Attention (GQA).
  - Uses basic `torch.matmul` operations instead of fused attention functions.
- **Updated `Attention.forward` Method**: Calls the new function with GQA parameters, ensuring shape alignment.
- **Why This Change?**: MPS has stricter tensor shape requirements than CUDA or CPU. The custom function avoids broadcasting issues that triggered `mps_matmul` errors, while preserving GQA functionality.

---

## Why This Works
- **Shape Compatibility**: Manually aligning tensor shapes eliminates MPS-specific errors.
- **Simplified Operations**: Relies on well-supported `torch.matmul`, avoiding MPS limitations.
- **No Functional Loss**: Maintains the original attention logic for consistent model behavior.

---

## Testing and Validation
- Tested on an M-series Mac with MPS enabled.
- Confirmed successful audio generation without errors.
- Verified GPU usage via macOS Activity Monitor during inference. (uses 99% of the GPU through out the inference)

---

## Recommendations for Future Maintenance
- **Update PyTorch Regularly**: MPS support improves with releases, potentially reducing custom workarounds.
- **Debugging Tip**: Log tensor shapes in attention computations for troubleshooting.
- **GQA Option**: Temporarily set `num_query_heads = num_kv_heads` in the config for debugging (may affect performance).

---

## Conclusion
This pull request makes the Dia 1.6B model fully functional on Apple Silicon GPUs by fixing MPS compatibility issues detailed in #197. The changes in `dia/layers.py` provide a robust, user-friendly solution. Reviewers are encouraged to test on their hardware and share feedback.

After committing these changes, I've manually tested the model by running it on:
1. Nvidia A100 (80GB VRAM) - RunPod
2. MacBook Pro M3 Pro (Using both GPU and CPU)

Here are the results
### Using Nvidia A100 GPU (80GB VRAM)

```bash
generate: starting generation loop
generate: using use_torch_compile=True, the first step may be slow
generate step 86: speed=6.716 tokens/s, realtime factor=0.078x
generate step 172: speed=218.047 tokens/s, realtime factor=2.535x
generate step 258: speed=218.079 tokens/s, realtime factor=2.536x
generate step 344: speed=217.908 tokens/s, realtime factor=2.534x
generate step 430: speed=217.129 tokens/s, realtime factor=2.525x
generate step 516: speed=217.717 tokens/s, realtime factor=2.532x
generate step 602: speed=217.898 tokens/s, realtime factor=2.534x
generate step 688: speed=217.757 tokens/s, realtime factor=2.532x
generate: avg steps=707.0, total duration=41.200s
```

### Using MPS (MacBook Pro M3 Pro - 36 GB RAM)

```bash
generate: starting generation loop
generate step 86: speed=7.759 tokens/s, realtime factor=0.090x
generate step 172: speed=8.470 tokens/s, realtime factor=0.098x
generate step 258: speed=8.536 tokens/s, realtime factor=0.099x
generate step 344: speed=8.489 tokens/s, realtime factor=0.099x
generate step 430: speed=8.607 tokens/s, realtime factor=0.100x
generate step 516: speed=8.615 tokens/s, realtime factor=0.100x
generate step 602: speed=8.592 tokens/s, realtime factor=0.100x
generate step 688: speed=8.597 tokens/s, realtime factor=0.100x
generate: avg steps=747.0, total duration=91.026s
```

### Using CPU (MacBook Pro M3 Pro - 36 GB RAM)

```bash
generate: starting generation loop
generate step 86: speed=2.531 tokens/s, realtime factor=0.029x
generate step 172: speed=2.639 tokens/s, realtime factor=0.031x
generate step 258: speed=2.619 tokens/s, realtime factor=0.030x
generate step 344: speed=2.645 tokens/s, realtime factor=0.031x
generate step 430: speed=2.647 tokens/s, realtime factor=0.031x
generate step 516: speed=2.577 tokens/s, realtime factor=0.030x
generate step 602: speed=2.603 tokens/s, realtime factor=0.030x
generate step 688: speed=2.626 tokens/s, realtime factor=0.031x
generate: avg steps=724.0, total duration=284.937s
```

### CODE REVIEW

**Comment by V12Hero** [comment]
@buttercrab I was able to make some modifications, and now the model runs on the GPUs of a MacBook Pro with an M-series chip. You guys might wanna take your time to review it. I've also updated the example. You can also view issue #197

File: dia/layers.py [hunk 1]
```diff
@@ -248,14 +301,14 @@ def forward(
             else:
                 attn_k, attn_v = cache.update(Xk_BxKxSxH, Xv_BxKxSxH, current_idx)
 
-        attn_output = F.scaled_dot_product_attention(
-            Xq_BxNxTxH,
-            attn_k,
-            attn_v,
+        attn_output = custom_scaled_dot_product_attention(
```
Comment (buttercrab) [inline dia/layers.py:304]: Can you only change when the backend is mps?

**Review by buttercrab — CHANGES_REQUESTED** [review]

File: dia/layers.py (hunk 1 above)
Comment (V12Hero) [reply dia/layers.py:304]: Alright, lemme see about it.

Comment (V12Hero) [reply dia/layers.py]: I've made the adjustments and tested it out on both MacBook and A100 (40GB VRAM this time from Colab), and it works fine, and each uses their respective dot product function, i.e., MPS uses `custom_scaled_dot_product_attention` and Cuda uses `F.scaled_dot_product_attention`.

**Review by buttercrab — APPROVED** [review]
LGTM!
